api/providers/views.py

```diff
@@ -792,3 +803,79 @@ def get_default_queryset(self):
 
     def get_queryset(self):
         return self.get_queryset_from_request()
+
+
+class RegistrationBulkCreate(APIView):
+    parser_classes = [FileUploadParser]
+
+    permission_classes = (
+        drf_permissions.IsAuthenticated,
+        base_permissions.TokenHasScope,
+        CanAddModerator,
+    )
+
+    def get_hash(self, file_obj):
+        BLOCK_SIZE = 65536
+        file_hash = hashlib.md5()
+        block = file_obj.read(BLOCK_SIZE)
+        while len(block) > 0:
+            file_hash.update(block)
+            block = file_obj.read(BLOCK_SIZE)
+        file_obj.seek(0)
+        return file_hash.hexdigest()
+
+    def put(self, request, *args, **kwargs):
```

Comment:
And, to be clear, if you don't call `check_object_permissions`, then your API view is not protected by the permissions classes you have above. Since there aren't any tests on this, you'll want to verify that the API is protected. And, of course, add tests before too long.

Reply:
There must be some magic in our permissions classes' `has_permission` function to make sure it's called automatically, because we're not explicitly calling `check_object_permissions` on our SchemaResponse code, but the permissions are definitely being checked correctly (source: ~120 tests per-method per-endpoint)

Reply:
I digged into the drf code a bit. It seems for `APIView`, `check_permission()` is invoked when the request is initialized (https://github.com/encode/django-rest-framework/blob/71e6c30034a1dd35a39ca74f86c371713e762c79/rest_framework/views.py#L415). This seems to check broad permissions (e.g. IsAuthenticated). However for row-level/object-level perms, we need to call `check_object_permissions()` ourselves. I'll add that to the start of the `put()` handler.

Reply:
Incidentally, this discussion helped me resolve an issue in my own API testing, so thanks, y'all

Reply:
@mfraezz Instead of adding `get_provider()` to the view, I can simply make this view inherit `ProviderMixin`, which already has `get_provider()` defined.